File: recursion/standard-uses/BST_classes.py
# Isaac Computer Science
# Usage licensed under the Open Government Licence v3.0

# In this example, a class is defined for a Node. However, it used here as a record structure (i.e. no methods are defined)


class Node:
    """Node for a binary search tree"""
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
    

class BST:
    """Binary search tree."""
    def __init__(self):
        self.root = None
        

    def get_root(self):
        """Returns value of root node."""
        return self.root
    

    def insert (self, data):
        """Insert new node."""
        new_node = Node(data)
        if self.root == None:  # New root node
            self.root = new_node
        else:                   
            placed = False
            current = self.root
            while placed == False:
                if current.data > data:  # Go left
                    if current.left == None:
                        current.left = new_node
                        placed = True
                    else:
                        current = current.left
                else:  # Go right
                    if current.right == None:
                        current.right = new_node
                        placed = True
                    else:
                        current = current.right                    
                    
                   
if __name__ == '__main__':
    # Just for testing
    bst = BST()
    bst.insert('fred')
    bst.insert('elaine')
    bst.insert('george')
    bst.insert('harry')     
        

File: recursion/standard-uses/BST_classes.cs
/*
Note: This file is designed to be copied out and compiled on your machine. 
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file. 

To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Compile the program
4. Run the program
*/

using System;

namespace Recursion
{
    
    // Node in a binary search tree (BST)
    class Node
    {
        
        public string data;
        public Node left;
        public Node right;

        public Node(string new_data) {
            data = new_data;
        }
        
    }

    // Binary search tree
    class BST
    {
        public Node root;

        public Node GetRoot() {
            return root;
        }
        

        // Insert a new node
        public void Insert(string data) {            
            Node newNode = new Node(data);
            if (root == null) {
                root = newNode;
            } else {
                bool placed = false;
                Node current = root;
                while (placed == false) {
                    if (current.data[0] > data[0]) {  // Go left
                        if (current.left == null) {
                            current.left = newNode;
                            placed = true;
                        } else {
                            current = current.left;
                        }
                    } else {  // Go right
                        if (current.right == null) {
                            current.right = newNode;
                            placed = true;
                        } else {
                            current = current.right;
                        }
                    }
                }
            }
        }
        
        
    }
}
